### PR TITLE
Update import order rules for Eslint

### PR DESCRIPTION
**Pull request by Antreesy**
### 🖼️ Screenshots

🏚️ Before | 🏡 After
---|---
.vue files | ---
image | image
.js files | ---
image | image
.spec.js files | ---
image | image

Changes in file on screenshot are to visualise groups separation, they are not a part of PR 

### 🚧 TODO

- [ ] Backport to stable24 / stable25 :warning: ONLY LINT RULES :warning: 
- [ ] Run lint:fix after merge as follow up

### 🏁 Checklist

- [X] ⛑️ Tests (unit and/or integration) are included
- [X] 📘 API documentation in `docs/` has been updated or is not required
- [X] 📗 User documentation in https://github.com/nextcloud/documentation/tree/master/user_manual/talk has been updated or is not required
- [X] 🔖 Capability is added or not needed

### CODE REVIEW

**Review by nickvergessen — APPROVED** [review]
Looks okay for me

**Comment by nickvergessen** [comment]
I think if we do something like this, we should do it directly before the stable26 branch of, to avoid having conflicts on too many backports

**Review by marcoambrosini — APPROVED** [review]

**Review by ShGKme — COMMENTED** [review]
First, I am a fan of code conventions with linting in general.

Specifically with imports order: most of the time, I don't see any imports at all (the imports block is collapsed in IDE), new imports are added by IDE with auto-import. Though, sometimes we look into imports and look on a file in web (GitHub) UI, git diff, and etc. So it may be useful in some cases and make the code just look better in general.

However, these rules affect almost all `.vue` components in the project and may lead to conflicts and issues on merges. For example, all current PR will have to be rebased.

Especially that could be a problem on backporting. Maybe we should backport eslint config and run lint fix before further backports.

File: .eslintrc.js [hunk 1]
```diff
@@ -9,5 +10,58 @@ module.exports = {
...
+								pattern: 'vue-material-design-icons/**',
+								group: 'internal',
+								position: 'after',
+							},
+							{
+								// group nextcloud-vue components after other @nextcloud imports
+								pattern: '@nextcloud/vue/dist/Components/**',
+								group: 'internal',
+								position: 'before',
+							},
```
[the remark this answers is not in the document]
Comment (Antreesy) [reply .eslintrc.js:40]: It makes sense for me to treat only components separately. In other case, we could omit this rule and treat all `nextcloud` imports together

File: .eslintrc.js [hunk 2]
```diff
@@ -9,5 +10,58 @@ module.exports = {
...
+							{
+								// group all project components after nextcloud-vue components
+								pattern: '*.vue',
+								patternOptions: { matchBase: true },
+								group: 'internal',
+								position: 'before',
+							},
+						],
+						pathGroupsExcludedImportTypes: ['vue-material-design-icons'],
+						'newlines-between': 'always',
```
[the remark this answers is not in the document]
Comment (Antreesy) [reply .eslintrc.js:56]: order in `pathGroups` matters for linter, and logical NOT is poorly supported. I could try fix this rule again, but as we don't have internal imports (which exist if we have aliases for folders), we could use this roundabout

File: .eslintrc.js [hunk 3]
```diff
@@ -9,5 +10,58 @@ module.exports = {
 				'node/no-unpublished-import': 0,
 			},
 		},
+		{
+			files: ['**/*.vue'],
+			rules: {
```
[the remark this answers is not in the document]
Comment (Antreesy) [reply .eslintrc.js:15]: I would vote for separate rules for .js and .spec.js, as most of the current configuration makes no sense for them

**Comment by nickvergessen** [comment]
We will no longer backport most of the things to stable25 once stable26 is there, so as long as we merge before 26 is split of, backporting should not be an issue.

The amount of open frontend PRs on split of should be limited anyway and when the version requirements are bumped rebasing is required anyway, so also not an issue or well a one time effort

**Comment by nickvergessen** [comment]
hmm so CI still passes, but there are warnings yielded all over it:

> /home/runner/work/spreed/spreed/src/App.vue
Warning:   41:1  warning  There should be at least one empty line between import groups                                                                                       import/order
Warning:   43:1  warning  There should be at least one empty line between import groups                                                                                       import/order
Warning:   44:1  warning  There should be at least one empty line between import groups                                                                                       import/order
Warning:   45:1  warning  There should be at least one empty line between import groups

**Comment by Antreesy** [comment]
> hmm so CI still passes, but there are warnings yielded all over it:

I didn't run linter, cause I think, if we're going to backport it, only lint rules should be backported. And after that run linter for each stable version (to not mess with non-existed components in old versions)

File: .eslintrc.js [hunk 4]
```diff
@@ -9,5 +9,58 @@ module.exports = {
 				'node/no-unpublished-import': 0,
 			},
 		},
+		{
+			files: ['**/*.js', '**/*.vue'],
```
Comment (ShGKme) [inline .eslintrc.js:13]: `files: ['**/*.js', '**/*.vue']` with current run script [eslint --ext .js,.vue src](https://github.com/nextcloud/spreed/blob/373c63e3ccfd03d845a7cf5ee4c7b63bcbf065cb/package.json#L14) means "all files". So after adding `.js` all these rules are not overrides but just rules.

```js
	rules: {
		'import/newline-after-import': 1,
		'import/order': [
			'warn',
			{
				groups: ['builtin', 'external', 'internal', ['parent', 'sibling', 'index'], 'unknown'],
				pathGroups: [
					{
						// group all style imports at the end
						pattern: '{*.css,*.scss}',
						patternOptions: { matchBase: true },
						group: 'unknown',
						position: 'after',
					},
					{
						// group material design icons
						pattern: 'vue-material-design-icons/**',
						group: 'external',
						position: 'after',
					},
					{
						// group @nextcloud imports
						pattern: '@nextcloud/{!(vue),!(vue)/**}',
						group: 'external',
						position: 'after',
					},
					{
						// group @nextcloud/vue imports
						pattern: '{@nextcloud/vue,@nextcloud/vue/**}',
						group: 'external',
						position: 'after',
					},
					{
						// group project components
						pattern: '*.vue',
						patternOptions: { matchBase: true },
						group: 'external',
						position: 'after',
					},
				],
				pathGroupsExcludedImportTypes: ['@nextcloud', 'vue-material-design-icons'],
				'newlines-between': 'always',
				alphabetize: {
					order: 'asc',
					caseInsensitive: true,
				},
				warnOnUnassignedImports: true,
			},
		],
	},

	overrides: [
		{
			files: ['**/*.spec.js'],
			rules: {
				'node/no-unpublished-import': 0,
			},
		},
	],
```

**Review by ShGKme — APPROVED** [review]

**Pull request merged**